### PR TITLE
Use a lock on cleanup, expire_tool and cache_tool

### LINKED ISSUE #7444: Prevent concurrent access to some ``ToolCache`` class attributes
Both `cleanup()` and `cache_tool()` methods of `ToolCache` access the following class attributes when a tool is reloaded:

- `_hash_by_tool_paths`
- `_mod_time_by_path`
- `_new_tool_ids`
- `_tools_by_path`
- `_tool_paths_by_id`

This causes the failure of unit test `ToolBoxTestCase.test_tool_reload_for_broken_tool`, see e.g.: https://travis-ci.org/galaxyproject/galaxy/jobs/476364519

### PR DESCRIPTION
**Pull request by mvdbeek**
This should fix https://github.com/galaxyproject/galaxy/issues/7444.

### CODE REVIEW

**Comment by nsoranzo** [comment]
Damn, at the third re-run it failed.

File: lib/galaxy/tools/cache.py [hunk 1]
```diff
@@ -30,22 +34,23 @@ def cleanup(self):
...
-                del self._tools_by_path[config_filename]
-                for tool_id in tool_ids:
-                    if tool_id in self._tool_paths_by_id:
-                        del self._tool_paths_by_id[tool_id]
-                removed_tool_ids.extend(tool_ids)
-            for tool_id in removed_tool_ids:
-                self._removed_tool_ids.add(tool_id)
-                if tool_id in self._new_tool_ids:
-                    self._new_tool_ids.remove(tool_id)
+            with self._lock:
```
Comment (nsoranzo) [inline lib/galaxy/tools/cache.py]: Should this be outside the try/except?

File: lib/galaxy/tools/cache.py [hunk 2]
```diff
@@ -79,31 +84,33 @@ def get_tool_by_id(self, tool_id):
...
+            self._new_tool_ids.add(tool_id)
+            for macro_path in tool._macro_paths:
+                self._mod_time_by_path[macro_path] = os.path.getmtime(macro_path)
+                if tool_id not in self._macro_paths_by_id:
+                    self._macro_paths_by_id[tool_id] = {macro_path}
+                else:
+                    self._macro_paths_by_id[tool_id].add(macro_path)
 
     def reset_status(self):
         """
```
Comment (nsoranzo) [inline lib/galaxy/tools/cache.py:122]: Not relevant for the unit test failure, but probably also this method should use the lock.

Comment (mvdbeek) [reply lib/galaxy/tools/cache.py:122]: Yep, was just looking at that too.

**Review by nsoranzo — APPROVED** [review]

File: lib/galaxy/tools/cache.py [hunk 3]
```diff
@@ -69,49 +80,54 @@ def _should_cleanup(self, config_filename):
 
     def get_tool(self, config_filename):
         """Get the tool at `config_filename` from the cache if the tool is up to date."""
-        return self._tools_by_path.get(config_filename, None)
+        with self._lock:
```
Comment (mvdbeek) [inline lib/galaxy/tools/cache.py]: I don't think this is necessary or a good idea, getting things from a dict is an atomic operation. I think we only need to lock add and remove (and I think that's why cleanup was already wrapped in try/except).

**Comment by mvdbeek** [comment]
So the circle ci log tells me that the guts of the logic is working fine, even if the assertion fails:
```
self = <unit.tools.test_toolbox.ToolBoxTestCase testMethod=test_tool_reload_for_broken_tool>

    def test_tool_reload_for_broken_tool(self):
        self._init_tool(filename="simple_tool.xml", version="1.0")
        self._add_config("""<toolbox><tool file="simple_tool.xml"/></toolbox>""")
        toolbox = self.toolbox
        tool = toolbox.get_tool('test_tool')
        assert tool is not None
        assert tool.version == "1.0"
        assert tool.tool_errors is None
        # Tool is loaded, now let's break it
        tool_path = tool.config_file
        with open(tool.config_file, 'w') as out:
            out.write('certainly not a valid tool')
    
        def check_tool_errors():
            tool = self.app.toolbox.get_tool("test_tool")
            assert tool is not None
            assert tool.version == "1.0"
            assert tool.tool_errors == 'Current on-disk tool is not valid'
    
        self._try_until_no_errors(check_tool_errors)
    
        # Tool is still loaded, lets restore it with a new version
        self._init_tool(filename="simple_tool.xml", version="2.0")
    
        def check_no_tool_errors():
            tool = self.app.toolbox.get_tool("test_tool")
            assert tool is not None
            assert tool.version == "2.0"
            assert tool.tool_errors is None
            assert tool_path == tool.config_file, "config file"
    
>       self._try_until_no_errors(check_no_tool_errors)

test/unit/tools/test_toolbox.py:236: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <unit.tools.test_toolbox.ToolBoxTestCase testMethod=test_tool_reload_for_broken_tool>
f = <function check_no_tool_errors at 0x7fb792404140>

    def _try_until_no_errors(self, f):
        e = None
        for i in range(2):
            try:
                f()
                return
            except AssertionError as error:
                e = error
                time.sleep(1)
    
>       raise e
E       AssertionError: assert '1.0' == '2.0'
E         - 1.0
E         + 2.0

test/unit/tools/test_toolbox.py:248: AssertionError
----------------------------- Captured stderr call -----------------------------
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:Created new ToolCache
DEBUG:galaxy.tools.toolbox.watcher:Starting ToolConfWatcher thread
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
INFO:galaxy.tools.toolbox.base:Parsing the tool configuration /tmp/tmplpgMM0/tool_conf.xml
DEBUG:galaxy.tools.cache:Caching tool
DEBUG:galaxy.tools.cache:Done caching tool
DEBUG:galaxy.tools.toolbox.base:Loaded tool id: test_tool, version: 1.0 into tool panel..
DEBUG:galaxy.tools.toolbox.base:Reading tools from config files finished (1.502 ms)
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Removed the following tools from cache: ['test_tool']
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
INFO:galaxy.tools.toolbox.base:Parsing the tool configuration /tmp/tmplpgMM0/tool_conf.xml
ERROR:galaxy.tools.toolbox.base:Tool '/tmp/tmplpgMM0/simple_tool.xml' is not valid:
Traceback (most recent call last):
  File "/home/circleci/repo/lib/galaxy/tools/toolbox/base.py", line 754, in load_tool
    tool = self.create_tool(config_file=config_file, tool_shed_repository=tool_shed_repository, guid=guid, **kwds)
  File "/home/circleci/repo/lib/galaxy/tools/__init__.py", line 272, in create_tool
    raise e
ParseError: syntax error: line 1, column 0
DEBUG:galaxy.tools.cache:Caching tool
DEBUG:galaxy.tools.cache:Done caching tool
DEBUG:galaxy.tools.toolbox.base:Reading tools from config files finished (0.741 ms)
DEBUG:galaxy.tools.toolbox.base:Loaded tool id: test_tool, version: 1.0 into tool panel..
DEBUG:galaxy.tools.toolbox.base:Loading tool panel finished (0.303 ms)
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Removed the following tools from cache: ['test_tool']
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
INFO:galaxy.tools.toolbox.base:Parsing the tool configuration /tmp/tmplpgMM0/tool_conf.xml
DEBUG:galaxy.tools.cache:Caching tool
DEBUG:galaxy.tools.cache:Done caching tool
DEBUG:galaxy.tools.toolbox.base:Reading tools from config files finished (1.566 ms)
DEBUG:galaxy.tools.toolbox.base:Loaded tool id: test_tool, version: 2.0 into tool panel..
DEBUG:galaxy.tools.toolbox.base:Loading tool panel finished (0.474 ms)
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
```
So I wondering if this is maybe a scope issue with the assertion function.

**Comment by mvdbeek** [comment]
OK, I think this is working now, I suspect https://github.com/galaxyproject/galaxy/pull/7446/commits/c3022c1c055d1ecba4213bdfb02f00a52194d928 fixed it, but the other commits surely make this more robust in general.

**Comment by nsoranzo** [comment]
@mvdbeek Fantastic, I'd have never find out what was the issue, thanks a bunch for looking into this!

Maybe we can drop the fsync now?

**Comment by nsoranzo** [comment]
Also you may want to rebase to get rid of the circleci commit from the PR that was just merged.

**Comment by mvdbeek** [comment]
Sure, let's try that.

**Review by nsoranzo — APPROVED** [review]

**Comment by nsoranzo** [comment]
@mvdbeek I've rebased again merging and cleaning the commits, since we have gone back and forth on some changes. Hope that's OK (and more readable in future).

**Comment by mvdbeek** [comment]
Sure, np.

**Pull request merged**